osquery/events/windows/usn_journal_reader.cpp

```diff
@@ -0,0 +1,893 @@
+
+bool GetEventString(std::string& buffer, const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  const wchar_t* filename = nullptr;
+  size_t name_length = 0U;
+  size_t record_length = 0U;
+
+  // NOTE(ww): I think this could be simplified: both V2 and V3 records
+  // contain a direct pointer to the filename that's guaranteed to be valid
+  // under the FSCTL_* constants we're using, so we might not need to muck
+  // with the offset at all.
+  switch (record->MajorVersion) {
+  case 2U: {
+    auto record_v2 = reinterpret_cast<const USN_RECORD_V2*>(record);
+
+    filename = record_v2->FileName;
+    name_length = static_cast<size_t>(record_v2->FileNameLength);
+    record_length = static_cast<size_t>(record_v2->RecordLength);
+
+    break;
+  }
+
+  case 3U: {
+    auto record_v3 = reinterpret_cast<const USN_RECORD_V3*>(record);
+
+    filename = record_v3->FileName;
+    name_length = static_cast<size_t>(record_v3->FileNameLength);
+    record_length = static_cast<size_t>(record_v3->RecordLength);
+
+    break;
+  }
+
+  default:
+    return false;
+  }
+
+  if (name_length == 0U) {
+    buffer.clear();
+    return true;
+  }
+
+  // Make sure we are not going outside of the record boundaries
+  auto record_start_ptr = reinterpret_cast<const uint8_t*>(record);
+  auto record_end_ptr = record_start_ptr + record_length;
+
+  auto string_end_ptr =
+      reinterpret_cast<const uint8_t*>(filename) + name_length;
+
+  if (string_end_ptr > record_end_ptr) {
+    LOG(ERROR) << "Invalid string length"
+               << "record size:" << record_length
+               << " name length: " << name_length;
+
+    return false;
+  }
+
+  std::wstring wide_chars_file_name(filename, (name_length / sizeof(wchar_t)));
+  buffer = wstringToString(wide_chars_file_name.c_str());
+
+  return true;
+}
+} // namespace USNParsers
+
+std::ostream& operator<<(std::ostream& stream,
+                         const USNJournalEventRecord::Type& type) {
+  auto it = kNTFSEventToStringMap.find(type);
+  if (it == kNTFSEventToStringMap.end()) {
+    stream << "UnknownEventRecordType";
+    return stream;
+  }
+
+  const auto& label = it->second;
+  stream << label;
+
+  return stream;
+}
+
+std::ostream& operator<<(std::ostream& stream,
+                         const USNFileReferenceNumber& ref) {
+  std::ios_base::fmtflags original_stream_settings(stream.flags());
+
+  FILE_ID_DESCRIPTOR native_file_id = {};
+  GetNativeFileIdFromUSNReference(native_file_id, ref);
+
+  stream << "0x";
+
+  switch (native_file_id.Type) {
+  case FileIdType: {
+    stream << std::hex << std::setfill('0') << std::setw(16)
+           << native_file_id.FileId.QuadPart;
+    break;
+  }
+  default: {
+    const auto& object_id = native_file_id.ObjectId;
+    stream << std::hex << std::setfill('0');
+    stream << std::setw(8) << object_id.Data1;
+    stream << std::setw(4) << object_id.Data2;
+    stream << std::setw(4) << object_id.Data3;
+    for (int i = 0; i < 8; ++i) {
```

Comment:
Why `int`?

Reply:
No particularly good reason -- this only gets used in debugging prints, but I can make it `auto` instead.